src/image/PhotoCalib.cc

```diff
@@ -0,0 +1,409 @@
+/*
+ * LSST Data Management System
+ * Copyright 2017 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include <cmath>
+
+#include "ndarray.h"
+#include "lsst/pex/exceptions.h"
+#include "lsst/afw/image/PhotoCalib.h"
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/math/BoundedField.h"
+#include "lsst/afw/table/io/OutputArchive.h"
+#include "lsst/afw/table/io/CatalogVector.h"
+#include "lsst/afw/table/Source.h"
+#include "lsst/daf/base/PropertySet.h"
+
+namespace lsst {
+namespace afw {
+namespace image {
+
+// ------------------- helpers -------------------
+
+namespace {
+
+double pow2(double value)
+{
+    return value*value;
+}
+
+double toMaggies(double counts, double scale) {
+    return counts/scale;
+}
+
+double toMagnitude(double counts, double scale) {
+    return -2.5*log10(counts/scale);
+}
+
+double toMaggiesSigma(double counts, double countsSigma, double scale, double scaleSigma, double maggies)
+{
+    return maggies*sqrt(pow2(countsSigma/counts) + pow2(scaleSigma/scale));
+}
+
+double toMagnitudeSigma(double counts, double countsSigma, double scale, double scaleSigma, double maggies)
+{
+    return 2.5*log(10.0) * sqrt(pow2(countsSigma/counts) + pow2(scaleSigma/scale));
+}
+
+void throwOnNegativeCounts(double counts, bool throwOnNegativeFlux) {
+    if (throwOnNegativeFlux)
+        if (counts < 0)
+        {
+            std::ostringstream os;
+            os << "Flux must be >= 0: saw " << counts;
+            throw LSST_EXCEPT(lsst::pex::exceptions::DomainError, os.str());
+        }
+}
+
+} // anonymous namespace
+
+// ------------------- Conversions to Maggies -------------------
+
+double PhotoCalib::countsToMaggies(double counts,
+                       const afw::geom::Point<double, 2> &point,
+                       bool throwOnNegativeFlux) const {
+    throwOnNegativeCounts(counts, throwOnNegativeFlux);
+    if (_isConstant)
+        return toMaggies(counts, _countsMag0);
```

Comment:
(I did forget to run clang-format before submitting for review: I'll fix these manually, and also run clang-format when I'm done with your review comments.)